otherlibs/systhreads/st_stubs.c

```diff
@@ -661,42 +683,39 @@ CAMLprim value caml_thread_new(value clos)
 /* the thread lock is not held when entering */
 CAMLexport int caml_c_thread_register(void)
 {
+  /* Systhreads initialized? */
+  if (!threads_initialized) return 0;
   /* Already registered? */
   if (This_thread != NULL) return 0;
 
+  /* At this point we should not hold any domain lock */
   CAMLassert(Caml_state_opt == NULL);
-  caml_init_domain_self(Dom_c_threads);
 
-  /* Take master lock to protect access to the runtime */
+  /* Acquire lock of domain */
+  caml_init_domain_self(Dom_c_threads);
   thread_lock_acquire(Dom_c_threads);
-  /* Create a thread info block */
-  caml_thread_t th = caml_thread_new_info();
-  /* If it fails, we release the lock and return an error. */
-  if (th == NULL) {
-    thread_lock_release(Dom_c_threads);
-    return 0;
-  }
-  /* Add thread info block to the list of threads */
-  CAMLassert(Active_thread != NULL);
-  th->next = Active_thread->next;
-  th->prev = Active_thread;
-  Active_thread->next->prev = th;
-  Active_thread->next = th;
 
-  /* Associate the thread descriptor with the thread */
-  st_tls_set(caml_thread_key, (void *) th);
-  /* Allocate the thread descriptor on the heap */
+  /* Create tick thread if not already done */
+  st_retcode err = create_tick_thread();
+  if (err != 0) goto out_err;
+
+  /* Set a thread info block */
+  caml_thread_t th = thread_alloc_and_add();
+  /* If it fails, we release the lock and return an error. */
+  if (th == NULL) goto out_err;
+  thread_init_current(th);
```

Comment:
I am missing something here. My reading of the code suggest that `thread_alloc_and_add` returns thread info block with an uninitialized runtime state (all fields are 0/NULL), and then `thread_init_current` calls `restore_runtime_state(th)` which "restores" this unitialized runtime state into `Caml_state`, resulting in a domain in an incorrect state.

I would naively expect a new thread to somehow get a valid `Caml_state` (this could be done by `caml_init_domain_self` for example), and then start by *saving* this valid runtime state in the thread info block.

Reply:
(Ah: maybe it is fine if the domain is in an invalid state, because we are giving up control so immediately another thread will be activated and restore its own (valid) state?)

Reply:
This is a bit counter-intuitive, but the thread info _is_ initialized with a valid runtime state indeed. I added a comment to clarify this in a new commit.

In reviewing this fact, I have found that I do not understand the initialization of `th->trap_barrier_off` which is in discrepancy with that in `domain_create`. This needs to be reviewed by someone knowledgeable before merging. (See the temporary `TODO` comment.)

(All this initialization between domain_create and caml_thread_new_info could be further factored, but I won't be doing this here.)

Sorry for piling up the commits, but upon the review you prompted, I fixed another issue with error handling in `caml_thread_new_info`, whereby the NULL return value of allocated memprof state was ignored.